src/Microsoft.Azure.SignalR.Common/ServiceConnections/ServiceConnectionBase.cs

```diff
@@ -406,13 +474,27 @@ private async Task<bool> ReceiveHandshakeResponseAsync(PipeReader input, Cancell
             }
         }
 
+        protected virtual void ProcessMessage(ServiceMessage message)
+        {
+            if (message is PingMessage ping && ping.TryGetValue(Constants.ShutdownState, out string val))
+            {
+                if (val == Constants.FinAck)
+                {
+                    _serviceConnectionRemoveTcs.TrySetResult(true);

```

Comment:
Why does this trigger the task? Shouldn't it just end the loop?

Reply:
No, we still have to wait for our `ASRS` to close the connection in a proper time.
What we can know from receiving a "FINACK" is no more new client connections will be assigned to this connection (Because our `ASRS` will remove this connection from its routing table).